### PR TITLE
Functional tests for pushing panels

### LINKED ISSUE #236: NUnit test for pushing data to GEM files
#### Definition of the test :

Continuation of IES toolkit functional tests as set out in #231, but testing pushing objects to GEM files.

Similar to the pull tests in the robot toolkit, a set of tests shall be created that tests:

- Pushing objects of type: Panel
    - with and without openings
    - with shades as 3D and 2D

For openings, a set of panels with a range of different openings will be created and pushed to GEM and then pulled to check if they pushed with the correct openings included, and then the same without any openings.
For shades, a set of panels that are of type `Shade` with openings will be created and pushed as 2D shades, and pulled, then the same with 3D shades, and repeat both tests but without openings.

### PR DESCRIPTION
**Pull request by Tom-Kingstone**
### Issues addressed by this PR

 Closes #236 

 
Functional tests for the the push functionality have been created that test the following:
- Pushing panels to GEM with 3D shades. Tests if the panels as they are in the json file are identical to the ones pushed and pulled, and checks that the correct number of spaces were created.
- Pushing panels to GEM with 2D shades with openings. Tests as above.

 ### Test files

 ### Changelog

 ### Additional comments

JSON files created in #237 used to push to a temporary file for testing.

### CODE REVIEW

**Comment by Tom-Kingstone** [comment]
@BHoMBot check compliance
@BHoMBot check required

File: .ci/unit-tests/IES_Adapter_Test/IES_Adapter_Test.csproj [hunk 1]
```diff
@@ -56,6 +56,11 @@
       <Private>True</Private>
       <SpecificVersion>False</SpecificVersion>
     </Reference>
+    <Reference Include="Data_Engine">
+      <HintPath>$(ProgramData)\BHoM\Assemblies\Data_Engine.dll</HintPath>
+		<Private>True</Private>
+		<SpecificVersion>False</SpecificVersion>
```
Comment (FraserGreenroyd) [inline .ci/unit-tests/IES_Adapter_Test/IES_Adapter_Test.csproj:62]:
```suggestion
      <HintPath>$(ProgramData)\BHoM\Assemblies\Data_Engine.dll</HintPath>
      <Private>True</Private>
      <SpecificVersion>False</SpecificVersion>
```

File: .ci/unit-tests/IES_Adapter_Test/PushTests.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            string currentDirectory = AppDomain.CurrentDomain.BaseDirectory;
+            List<string> paths = currentDirectory.Split('\\').ToList();
+            paths = paths.Take(paths.IndexOf(".ci") + 2).ToList();
+            string ModelsPath = Path.Join(string.Join("\\", paths), "Models");
+            FileSettings file = new FileSettings()
+            {
+                Directory = ModelsPath,
+                FileName = "PushedModel.gem"
+            };
+            m_Adapter = new IESAdapter();
```
Comment (FraserGreenroyd) [inline .ci/unit-tests/IES_Adapter_Test/PushTests.cs:61]:
```suggestion

            m_Adapter = new IESAdapter();
```

File: .ci/unit-tests/IES_Adapter_Test/PushTests.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            m_PushConfig = new PushConfigIES()
```
Comment (FraserGreenroyd) [inline .ci/unit-tests/IES_Adapter_Test/PushTests.cs:62]:
```suggestion

            m_PushConfig = new PushConfigIES()
```

File: .ci/unit-tests/IES_Adapter_Test/PushTests.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            {
+                PlanarTolerance = 1.0e-6,
+                AngleTolerance = Tolerance.Angle,
+                DecimalPlaces = 6,
+                DistanceTolerance = Tolerance.MacroDistance,
+                File = file
+            };
+            m_PullConfig = new PullConfigIES()
```
Comment (FraserGreenroyd) [inline .ci/unit-tests/IES_Adapter_Test/PushTests.cs:70]:
```suggestion

            m_PullConfig = new PullConfigIES()
```

File: .ci/unit-tests/IES_Adapter_Test/PushTests.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+            List<Panel> panels = Engine.Adapters.File.Compute.ReadFromJsonFile(Path.Combine(m_PullConfig.File.Directory, "Test Model 3D Shades.json"), true).Where(x => x.GetType() == typeof(Panel)).Cast<Panel>().ToList();
+
+            int count = 0;
+            foreach (Panel panel in panels)
+            {
+                panel.Name = count.ToString();
+                count++;
+            }
+
+
```
Comment (FraserGreenroyd) [inline .ci/unit-tests/IES_Adapter_Test/PushTests.cs:118]:
```suggestion
```

File: .ci/unit-tests/IES_Adapter_Test/PushTests.cs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            m_Adapter.Push(panels, actionConfig: m_PushConfig);
+
+            List<IBHoMObject> objs = m_Adapter.Pull(request, actionConfig:m_PullConfig).Cast<IBHoMObject>().ToList();
+
+            List<Space> pulledSpaces = BH.Engine.Environment.Query.Spaces(objs);
+            List<Panel> pulledPanels = BH.Engine.Environment.Query.Panels(objs);
+
+            pulledPanels = BH.Engine.Data.Query.OrderBy(pulledPanels, "Name");
+
+
```
Comment (FraserGreenroyd) [inline .ci/unit-tests/IES_Adapter_Test/PushTests.cs:129]:
```suggestion
```

File: .ci/unit-tests/IES_Adapter_Test/PushTests.cs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+            List<Panel> panels = Engine.Adapters.File.Compute.ReadFromJsonFile(Path.Combine(m_PullConfig.File.Directory, "Test Model 2D Shades.json"), true).Where(x => x.GetType() == typeof(Panel)).Cast<Panel>().ToList();
+
+            int count = 0;
+            foreach (Panel panel in panels)
+            {
+                panel.Name = count.ToString();
+                count++;
+            }
+
+
```
Comment (FraserGreenroyd) [inline .ci/unit-tests/IES_Adapter_Test/PushTests.cs:159]:
```suggestion
```

File: .ci/unit-tests/IES_Adapter_Test/PushTests.cs [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+            m_Adapter.Push(panels, actionConfig: m_PushConfig);
+
+            List<IBHoMObject> objs = m_Adapter.Pull(request, actionConfig: m_PullConfig).Cast<IBHoMObject>().ToList();
+
+            List<Space> pulledSpaces = BH.Engine.Environment.Query.Spaces(objs);
+            List<Panel> pulledPanels = BH.Engine.Environment.Query.Panels(objs);
+
+            pulledPanels = BH.Engine.Data.Query.OrderBy(pulledPanels, "Name");
+
+
```
Comment (FraserGreenroyd) [inline .ci/unit-tests/IES_Adapter_Test/PushTests.cs:170]:
```suggestion
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Review by FraserGreenroyd — APPROVED** [review]
Looks good now, thanks @Tom-Kingstone

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance
@BHoMBot check unit-tests

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: core, installer, null-handling, ready-to-merge, serialisation, versioning

**Pull request merged**